Source/WebCore/dom/FragmentDirectiveGenerator.cpp

```diff
@@ -0,0 +1,147 @@
+        if (potentialPreviousPosition.deepEquivalent().containerNode() != startPosition.deepEquivalent().containerNode())
+            break;
+        previousPosition = potentialPreviousPosition;
+    }
+
+    auto document = startPosition.deepEquivalent().document();
+    if (!document)
+        return { };
+
+    auto range = Range::create(*document);
+    RefPtr startNode = previousPosition.deepEquivalent().containerNode();
+    range->setStart(startNode.releaseNonNull(), previousPosition.deepEquivalent().computeOffsetInContainerNode());
+    RefPtr endNode = startPosition.deepEquivalent().containerNode();
+    range->setEnd(endNode.releaseNonNull(), startPosition.deepEquivalent().computeOffsetInContainerNode());
+
+    return range->toString().trim(isHTMLSpaceButNotLineBreak);
+}
+
+static String nextWordsFromPosition(uint numberOfWords, VisiblePosition& startPosition)
+{
+    auto nextPosition = startPosition;
+    while (numberOfWords--) {
+        auto potentialNextPosition = nextWordPosition(nextPosition);
+        if (potentialNextPosition.deepEquivalent().containerNode() != startPosition.deepEquivalent().containerNode())
+            break;
+        nextPosition = potentialNextPosition;
+    }
+
+    auto document = nextPosition.deepEquivalent().document();
+    if (!document)
+        return { };
+
+    auto range = Range::create(*document);
+    RefPtr startNode = startPosition.deepEquivalent().containerNode();
+    range->setStart(startNode.releaseNonNull(), startPosition.deepEquivalent().computeOffsetInContainerNode());
+    RefPtr endNode = nextPosition.deepEquivalent().containerNode();
+    range->setEnd(endNode.releaseNonNull(), nextPosition.deepEquivalent().computeOffsetInContainerNode());
+
+    return range->toString().trim(isHTMLSpaceButNotLineBreak);
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#generating-text-fragment-directives
+void FragmentDirectiveGenerator::generateFragmentDirective(const SimpleRange& textFragmentRange)
+{
+    LOG_WITH_STREAM(TextFragment, stream << " generateFragmentDirective: ");
+
+    String textDirectivePrefix = ":~:text="_s;
+
+    auto url = textFragmentRange.startContainer().document().url();
+
+    auto textFromRange = createLiveRange(textFragmentRange)->toString();
+
+    VisiblePosition visibleEndPosition = VisiblePosition(Position(textFragmentRange.protectedEndContainer(), textFragmentRange.endOffset(), Position::PositionIsOffsetInAnchor));
+    VisiblePosition visibleStartPosition = VisiblePosition(Position(textFragmentRange.protectedStartContainer(), textFragmentRange.startOffset(), Position::PositionIsOffsetInAnchor));
+
+    if (textFromRange.length() >= maxCharacters) {
+        String startText = percentEncodeFragmentDirectiveSpecialCharacters(nextWordsFromPosition(3, visibleStartPosition));
+        String endText = percentEncodeFragmentDirectiveSpecialCharacters(previousWordsFromPosition(3, visibleEndPosition));
```

Comment:
We need to think about what happens with Chinese / Japanese where by default every character as a distinct word.